datafusion/src/physical_plan/sort_preserving_merge.rs

```diff
@@ -246,7 +274,19 @@ impl SortKeyCursor {
             .zip(other.columns.iter())
             .zip(options.iter());
 
-        for ((l, r), sort_options) in zipped {
+        // Recall or initialise a collection of comparators for comparing
+        // columnar arrays of this cursor and "other".
+        let cmp = self
+            .batch_comparators
+            .entry(other.batch_idx)
+            .or_insert_with(|| Vec::with_capacity(other.columns.len()));
```

Comment:
Don't we need to clean older `batch_idx` over time?

Reply:
Do you mean to reduce the peak memory usage by freeing comparators?

Reply:
The HashMap only lives for as long as the lifetime of the sort preserving merge operator. I suppose you could merge n streams large enough to make the old comparators take up non-negligible memory.

Reply:
Yes, that's what I meant. The memory usage would increase over time by holding onto the older comparators.

Reply:
I personally think the "right" thing to do in this case is to create an arrow comparator interface that doesn't have bound arrays, so like
```
let comparator = arrow::compute::build_comparator(array1.data_type(), array2.data_type());

...

if comparator(array1, index_1, array2, index_2) == Ordering::Equal {
  ...
}
```

With that interface we could simply use the same comparator 

So that is to say, I recommend leaving this PR the way it is and putting our efforts into a better comparison interface rather than trying to optimize for a small amount of memory savings here